src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,90 @@
 
 import com.endava.internship.domain.Privilege;
 import com.endava.internship.domain.User;
-import com.endava.internship.service.UserService;
-
+import org.apache.commons.lang3.tuple.Pair;
 import java.util.List;
-import java.util.Map;
 import java.util.Optional;
+import java.util.Map;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().map(u -> u.getFirstName()).sorted((String u1, String u2) -> (u2.compareTo(u1))).collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        List<User> list = users.stream().sorted((User u1,User u2) ->(u1.getFirstName().compareTo(u2.getFirstName()))).sorted((User u1, User u2) -> u2.getAge().
+                compareTo(u1.getAge())).collect(Collectors.toList());
+        return list;
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        List<List<Privilege>> list = users.stream().map(x -> x.getPrivileges()).collect(Collectors.toList());
+        List<Privilege> list1 = list.stream().flatMap(x -> x.stream().distinct()).distinct().collect(Collectors.toList());
+        return list1;
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        Optional<User> user = users.stream().filter(x->x.getAge() > age).filter(u -> u.getPrivileges().contains(Privilege.UPDATE)).findFirst();
+        return user;
     }
```

Comment:
we don't need local variable user

instead filter(condition 1).filter(condition 2)
you can use filter(condition 1 && condition 2), and instead 'x', 'u' you could use `user`

Reply:
resolved